cmd/kube-rbac-proxy/app/options/proxyoptions.go

```diff
@@ -125,13 +140,87 @@ func (o *ProxyOptions) ApplyTo(c *server.KubeRBACProxyInfo, a *serverconfig.Auth
 		}
 	}
 
+	c.UpstreamHeaders = o.UpstreamHeader
 	c.IgnorePaths = o.IgnorePaths
 	c.AllowPaths = o.AllowPaths
 	a.APIAudiences = o.TokenAudiences
 
 	return nil
 }
 
+func validateSecureConnection(o *ProxyOptions) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(o.UpstreamDNSTimeout)*time.Second)
+	defer cancel()
+
+	isLoopback, err := isLoopbackAddress(ctx, o.Upstream)
+	if err == nil && isLoopback {
+		return nil
+	}
+
+	if o.UpstreamForceH2C {
+		return fmt.Errorf("loopback address is required for h2c")
+	}
+
+	// isMTLSPossible emphasizes that we can't verify that upstream verifies the
+	// client's certificate.
+
+	if err := validateMTLSConfiguration(o.UpstreamClientCertFile, o.UpstreamClientKeyFile, o.UpstreamCAFile); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateMTLSConfiguration(upstreamClientCertPath, upstreamClientKeyPath, upstreamCAPath string) error {
+	// Check if client is configured to provide a certificiate.
+	if len(upstreamClientCertPath) == 0 || len(upstreamClientKeyPath) == 0 || len(upstreamCAPath) == 0 {
+		return fmt.Errorf("krp client cert/key and an upstream CA are required for mTLS")
+	}
+
+	_, err := tls.LoadX509KeyPair(upstreamClientCertPath, upstreamClientKeyPath)
+	if err != nil {
+		return fmt.Errorf("failed to read upstream client cert/key: %w", err)
```

Comment:
the above does not have to be a read error, see https://cs.opensource.google/go/go/+/refs/tags/go1.22.1:src/crypto/tls/tls.go;l=245

Reply:
If we are able to read the file, but we can't interpret the content properly, it is still a read error.

I can change the wording to be more generic.